db/index.go

```diff
@@ -0,0 +1,495 @@
+	key, err := i.getDocKey(doc)
+	if err != nil {
+		return err
+	}
+	err = txn.Datastore().Put(ctx, key.ToDS(), []byte{})
+	if err != nil {
+		return NewErrFailedToStoreIndexedField(key.IndexID, err)
+	}
+	return nil
+}
+
+func (i *collectionSimpleIndex) Update(
+	ctx context.Context,
+	txn datastore.Txn,
+	oldDoc *client.Document,
+	newDoc *client.Document,
+) error {
+	key, err := i.getDocKey(oldDoc)
+	if err != nil {
+		return err
+	}
+	err = txn.Datastore().Delete(ctx, key.ToDS())
+	if err != nil {
+		return err
+	}
+	return i.Save(ctx, txn, newDoc)
+}
+
+func (i *collectionSimpleIndex) RemoveAll(ctx context.Context, txn datastore.Txn) error {
+	prefixKey := core.IndexDataStoreKey{}
+	prefixKey.CollectionID = strconv.Itoa(int(i.collection.ID()))
+	prefixKey.IndexID = strconv.Itoa(int(i.desc.ID))
+	q, err := txn.Datastore().Query(ctx, query.Query{
+		Prefix: prefixKey.ToString(),
+	})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := q.Close(); err != nil {
+			log.ErrorE(ctx, "Failed to close collection query", err)
+		}
+	}()
+
+	for res := range q.Next() {
+		if res.Error != nil {
+			return res.Error
+		}
+		err = txn.Datastore().Delete(ctx, ds.NewKey(res.Key))
+		if err != nil {
+			return NewCanNotDeleteIndexedField(err)
+		}
+	}
+
+	return nil
+}
+
+func (i *collectionSimpleIndex) Name() string {
+	return i.desc.Name
+}
+
+func (i *collectionSimpleIndex) Description() client.IndexDescription {
+	return i.desc
+}
+
+func validateIndexDescription(desc client.IndexDescription) error {
+	if desc.ID != 0 {
+		return NewErrNonZeroIndexIDProvided(desc.ID)
+	}
+	if len(desc.Fields) == 0 {
+		return ErrIndexMissingFields
+	}
+	if len(desc.Fields) == 1 && desc.Fields[0].Direction == client.Descending {
+		return ErrIndexSingleFieldWrongDirection
+	}
+	for i := range desc.Fields {
+		if desc.Fields[i].Name == "" {
+			return ErrIndexFieldMissingName
+		}
+		if desc.Fields[i].Direction == "" {
+			desc.Fields[i].Direction = client.Ascending
+		}
+	}
+	return nil
+}
+
+func generateIndexName(col client.Collection, fields []client.IndexedFieldDescription, inc int) string {
+	sb := strings.Builder{}
+	direction := "ASC"
+	//if fields[0].Direction == client.Descending {
+	//direction = "DESC"
+	//}
+	sb.WriteString(col.Name())
+	sb.WriteByte('_')
+	sb.WriteString(fields[0].Name)
+	sb.WriteByte('_')
+	sb.WriteString(direction)
+	if inc > 1 {
+		sb.WriteByte('_')
+		sb.WriteString(strconv.Itoa(inc))
+	}
+	return sb.String()
+}
+
+func (c *collection) CreateIndex(
+	ctx context.Context,
+	desc client.IndexDescription,
+) (client.IndexDescription, error) {
+	index, err := c.createIndex(ctx, desc)
+	if err != nil {
+		return client.IndexDescription{}, err
+	}
+	if c.isIndexCached {
+		c.indexes = append(c.indexes, index)
+	}
+	return index.Description(), nil
+}
+
+func (c *collection) DropIndex(ctx context.Context, indexName string) error {
+	key := core.NewCollectionIndexKey(c.Name(), indexName)
+
+	txn, err := c.getTxn(ctx, false)
```

Comment:
todo: This appears to have the same problem as `createIndex` - it doesn't look like you are commiting implicit transactions?